components/vision.py

```diff
@@ -0,0 +1,200 @@
+            return
+
+        # if we have already processed these results
+        timestamp = results.getTimestamp()
+        if timestamp == self.last_timestamp and wpilib.RobotBase.isReal():
+            return
+        self.last_timestamp = timestamp
+
+        # TODO Figure out the latency issue
+        # if abs(wpilib.Timer.getFPGATimestamp() - timestamp) > 0.5:
+        #    return # IT FAILS HERE
+
+        if results.multiTagResult.estimatedPose.isPresent:
+            p = results.multiTagResult.estimatedPose
+            pose, reprojectionErr = choose_pose_multi(
+                p, self.camera_to_robot, self.chassis.get_pose()
+            )
+
+            self.field_pos_obj.setPose(pose)
+
+            if self.add_to_estimator:
+                self.chassis.estimator.addVisionMeasurement(
+                    pose,
+                    timestamp,
+                    (reprojectionErr, reprojectionErr, reprojectionErr / 3),
+                )
+
+            if self.should_log:
+                ground_truth_pose = self.chassis.get_pose()
+                trans_error1: float = ground_truth_pose.translation().distance(
+                    p.best.translation()
+                )
+                trans_error2: float = ground_truth_pose.translation().distance(
+                    p.alt.translation()
+                )
+                rot_error1: float = (  # type: ignore
+                    ground_truth_pose.rotation() - p.best.rotation()
+                ).radians()
+                rot_error2: float = (  # type: ignore
+                    ground_truth_pose.rotation() - p.alt.rotation()
+                ).radians()
+
+                self.pose_log_entry.append(
+                    [
+                        p.best.x,
+                        p.best.y,
+                        typing.cast(float, p.best.rotation().radians()),
+                        trans_error1,  # error of main pose
+                        rot_error1,
+                        p.alt.x,
+                        p.alt.y,
+                        typing.cast(float, p.alt.rotation().radians()),
+                        trans_error2,
+                        rot_error2,
+                        ground_truth_pose.x,
+                        ground_truth_pose.y,
+                    ]
+                )
+
+        for target in results.getTargets():
+            poses = estimate_poses_from_apriltag(self.camera_to_robot, target)
+            if poses is None:
+                # tag doesn't exist
+                continue
+
+            best, alt, self.last_pose_z = poses
+            pose = choose_pose(
+                best,
+                alt,
+                self.chassis.get_pose(),
+            )
+
+            # filter out likely bad targets
+            if target.getPoseAmbiguity() > 0.25:
+                continue
+
+            self.field_pos_obj.setPose(pose)
+            self.chassis.estimator.addVisionMeasurement(pose, results.getTimestamp())
```

Comment:
As an aside the default values for the standard deviations on the chassis has an infinity for rotation which is certainly not right.

Reply:
Last year we had an infinity for the vision sd for rotation because we would like to think that the navx gyro on the robot is highly accurate so we didn't take in the vision measurement for rotation.

Reply:
Even if the navx is much more accurate there is no way it is infinitely more accurate. I think that our default should be a large but finite number.